components/region_cache_memory_engine/src/engine.rs

```diff
@@ -23,19 +35,86 @@ pub struct RegionMemoryEngine {
     data: [Arc<Skiplist<ByteWiseComparator>>; 3],
 }
 
+impl RegionMemoryEngine {
+    pub fn with_capacity(arena_size: usize) -> Self {
+        RegionMemoryEngine {
+            data: [
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    arena_size,
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
+impl Default for RegionMemoryEngine {
+    fn default() -> Self {
+        RegionMemoryEngine {
+            data: [
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+                Arc::new(Skiplist::with_capacity(
+                    ByteWiseComparator::default(),
+                    1 << 20,
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
 impl Debug for RegionMemoryEngine {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        unimplemented!()
+        write!(f, "Region Memory Engine")
     }
 }
 
-type SnapshotList = BTreeMap<u64, u64>;
+#[derive(Default)]
+struct SnapshotList(BTreeMap<u64, u64>);
+
+impl SnapshotList {
+    fn new_snapshot(&mut self, read_ts: u64) {
+        // snapshot with this ts may be granted before
+        let count = self.0.get(&read_ts).unwrap_or(&0) + 1;
+        self.0.insert(read_ts, count);
+    }
+
+    fn reclaim_snapshot(&mut self, read_ts: u64) {
```

Comment:
I think reclaim is a bit misleading name, as it means re-use the snapshot. 
Maybe remove_snapshot is better.

Reply:
Ok